20240919-scheduler-failure-policy.md

```diff
@@ -0,0 +1,135 @@
+
+### Future Design
+
+Although not part of the proposal, in future we can extend Scheduler to include a staging queue which is dedicated for Jobs where no current stream implements its target.
+This addition would mean that the Jobs with no target are not needlessly being attempted to be triggered, freeing up main queue resources and preserving the intended failure policy.
+Upon trigger, if no target stream is connected, the job will be moved to this staging queue.
+In the event a stream implementing the target is connected, the Job can be moved to the main queue and triggered immediately or on its proper schedule.
+
+### go-etcd-cron
+
+To support the new `FailurePolicy` options, we first need to keep track of the current number of attempts the current Job count has been triggered on a particular tick.
+We also need to keep track of the time at which that attempt was made in order to calculate the next attempt time according to the `FailurePolicy` schedule.
+Like the trigger time, the last attempt time is the virtual _correct_ time that the attempt was to be made, rather than the observed wall clock time.
+This is to ensure durability of the Scheduler during events of slow down or down time/restarts.
+The attempts and last attempt time will added to the Counter proto message, and managed by the trigger queue manager.
+Attempts will be 0 and last attempted time will be `nil` for all new tick counts of a Job.
+
+```proto
+// Counter holds counter information for a given job.
+message Counter {
+  ...
+
+  // attempts is the number of times the job has been attempted to be
+  // triggered at this count.
+  uint32 attempts = 4;
+
+  // last_attempted is the timestamp the job was last attempted to be triggered.
+  optional google.protobuf.Timestamp last_attempted = 5;
+}
+```
+
+Below are the proto definitions for the new `FailurePolicy` options.
+Both constant and cron policies include an optional max retries option to limit the number of retries according to the number of attempts.
+If max retries is unset, the Job will be retried indefinitely.
+The failure policy message is added as an optional field to the Job message.
+If unset, the failure policy of a Job is `Drop`.
+
+```proto
+// FailurePolicy defines the policy to apply when a job fails to trigger.
+message FailurePolicy {
+  // policy is the policy to apply when a job fails to trigger.
+  oneof policy {
+    FailurePolicyDrop drop = 1;
+    FailurePolicyConstant constant = 2;
+    FailurePolicyCron cron = 3;
+  }
+}
+
+// FailurePolicyDrop is a policy which drops the job tick when the job fails to
+// trigger.
+message FailurePolicyDrop {}
+
+// FailurePolicyRetry is a policy which retries the job at a consistent
+// delay when the job fails to trigger.
+message FailurePolicyConstant {
```

Comment:
This is redundant to the Cron schedule since you can just create a schedule with "every @1s"

Reply:
I like having a strongly typed interval as it means we don’t need to re-parse the value at every operation. I can foresee “@every 1s” being converted to the constant type as an optimisation.